### ISSUE TITLE
Feature request: option to generate migrations for trigger creation

### ISSUE THREAD

**Issue by simonw** [Opus10/django-pgtrigger#25]
It looks like `django-pgtrigger` currently works by creating or replacing any missing triggers when the app starts up (correct me if I'm wrong) rather than creating triggers using explicit migration scripts.

This makes me nervous, because of the way it could interact with historical data migrations.

My projects tend to make extensive use of data migrations. I often have a migration that creates a table, then a separate migration that populates that table with a core set of key records - for things like feature flags or other configuration-style data.

When I run `./manage.py migrate` in a new environment it may churn through dozens of migrations to create tables, populate them with data, update that data later in the migration sequence and so on.

Here's the problem: if any of those rows are affected by triggers (for denormalized columns for example) I need to execute the historic version of that trigger for that point in the migration history - NOT the most recent version if the trigger which may be incompatible with that historic table layout!

It's not mentioned in the documentation and an initial review of the code didn't spot anything, hence I'm classing this as a feature request: I would really appreciate the option to use this package in a way that can code generate SQL for creating triggers that I can include in a standard Django migration.

Even better: a management command that generates the appropriate migration file for me!

I've only just started looking at this project so it's very possible there is historic context I'm missing about this problem.

**Comment by simonw** [comment]
it looks like all of the pieces are there for me to implement this myself for my project: I can set `PGTRIGGER_INSTALL_ON_MIGRATE = False` and then reuse the code here to generate the SQL needed to install a trigger using a migration: 
https://github.com/jyveapp/django-pgtrigger/blob/588bfcc650e7082f9158582e4b0375ac0b8b9b4b/pgtrigger/core.py#L644-L658

**Comment by dracos** [comment]
"It looks like django-pgtrigger currently works by creating or replacing any missing triggers when the app starts up (correct me if I'm wrong)" - they are created in `post_migrate`, or manually via the management commands. So for a new environment, it is "okay" in terms of not having an invalid trigger halfway through, but not okay if you require the trigger along the way to fill in the data, as you say.

This kind of thing was previously mentioned in this ticket: https://github.com/Opus10/django-pgtrigger/issues/7 which was said to be fixed by the management command to disable/reenable individual triggers. But I don't think those really do solve it; I can't see us in production being able to disable a live trigger for a length of time.

I agree, ideally, when makemigrations is called as well as the extra Event models being added, the relevant trigger should be added then, rather than in `post_migrate`. I have this function:
```python
def add_triggers(apps, schema_editor):
    from pgtrigger.core import registry
    app_label = 'app'
    models = ['ModelA', 'ModelB']
    for uri in registry:
        for model in models:
            if uri.startswith(f"{app_label}.{model}:"):
                cls = apps.get_model(app_label, model)
                trigger = registry[uri][1]
                trigger.install(cls)
```

Which I *think* might work, added to any migration that needs to update a trigger? Not ideal having to manually think when it's needed, but might deal with the issue if it works.

**Comment by wesleykendall** [comment]
@dracos @simonw

By default, all triggers are installed/updated/deleted at the end of migrations in a large syncing operation. As mentioned, this can be turned off via `settings.PGTRIGGER_INSTALL_ON_MIGRATE` You can also use pgtrigger management commands to install or update triggers instead of going into library code. This, however, does not get around the problem you brought up @simonw of having the *historical* triggers installed and running, which is an obvious problem. And yes @dracos, disabling a trigger was just a bandaid to that particular problem.

## Triggers in migrations

I agree that the optimal approach for installing triggers is to make it part of the Django migration system. When I originally created this library, I wanted triggers to be declared in the model Meta (I couldn't find a way to extend Meta) and work with migrations automatically. Here are some of the problems I've faced:

1. I found no easy way of hooking into the migration system. While it's pretty easy to auto add migrations for new custom fields and stuff, I never found the ability to hook into it in a completely custom way for triggers so that it just works automatically as you change your trigger definitions. I should revisit this again now that the library is in a more polished state. If you know of any apps that do this outside of normal fields, indices, and DB extensions, do let me know so I can see how they do it. Triggers seem to be a bit of a different beast that might require a monkeypatch or other means to work with the migration system.

2. It is common to attach triggers to the Django User model and other third-party models, especially if you are using django-pghistory. Needing to integrate with third-party models also made an integration with the migration system a bit unclear. I register my third-party triggers in an apps.py in my project, so would migrations be placed inside of the app with the apps.py?

For 1 - I believe it is possible to have a pgtrigger command that manually makes data migrations, and ideally these data migrations have both forwards/backwards code that is a snapshot of the raw SQL of the trigger at the time. This would make it much easier to address https://github.com/Opus10/django-pgtrigger/issues/29 too. If there is the ability to integrate with makemigrations and do this automatically, that would be great, otherwise I will probably need to at least print a warning at the end of post_migrate that a user needs to create trigger migrations. Open to ideas, but there are a lot of unknowns before being definitively able to say which interface is best.

For 2 - I will have to check, but I believe it should be possible to determine which app a trigger registers itself in so that I can make the migrations in the proper directory when overriding third-party models. Either this or I could make users define a proxy model to add a trigger to a third-party model. If anyone knows of ways to directly integrate with Meta, that would be great too, but I remember it being an unexpectedly long rabbit hole the first time I tried.

Another weird edge case related to order of operations - if a trigger has a condition on model fields, and the model fields are dropped in a migration before the trigger is updated, I believe that this will just result in an outright error (and same for vice versa). Although possible to implement, it's edge cases like these that make it difficult to have something that just works. I think the ideal solution will probably end up being something that is a little more on the manual side for now and for the users to know how to resolve errors if things are installed out of order

## Next Steps

I think this is a big enough issue for new and potential users that I (or @tomage) should prioritize how to best add this in. This is something I won't even be able to look at until Aug 14, but I put it on my calendar that weekend to set aside time to really investigate the possibilities. If you all have any suggestions, do let me know! @tomage and I are also going to be better about more frequently checking and responding to issues weekly too. We recently migrated this project to a different open source org so that it was no longer tied to a company. We are always welcome to other contributors and maintainers at Opus10 if there are others that are really into triggers like we are :) (We are working on a MySQL version too)

**Comment by wesleykendall** [comment]
I looked into the possibility of integrating django-pgtrigger directly into the migration system. It is doable, although a bit hacky as expected. I will likely have to patch the core migration loader. I was able to get an example version of code working that just created `operations.RawSQL` migration operations for installing the entire raw SQL that ``django-pgtrigger`` typically performs behind the scenes. I still need to evaluate how I will be able to query previous migrations to get the raw SQL of previous trigger declarations. This will allow me to more easily make reversible `RawSQL` operations. Should be doable though, just need a bit more time to explore.

Although I think I can make things work most of the time, there are particular scenarios  that might be tricky to always get right. For example, if a user adds columns and references them in a trigger, we should create/update the trigger in the migration file at the end. I can't think of any examples where the trigger needs to be updated before the operation, but my hunch is that there are some options

Since ordering of the trigger installation is dependent on other operations, one easy way to alleviate these issues would be to drop/recreate the trigger in the migration file by default by surrounding all operations:

```
[
    migrations.RawSQL('drop trigger...', reverse_sql='create trigger with OLD definition'),
    migrations.RemoveField(
       model_name='journal',
       name='action',
    ),
    migrations.RawSQL('create trigger with new definition', reverse_sql='drop trigger')
]
```

If I always surround the migration operations with a drop/create for trigger changes, it seems to alleviate most issues that happen as a result of columns, table names, and other things changing.

Since Django makes migration files atomic by default, all I need to do is ensure that the initial trigger operation will take out a lock on the table so that no updates happen while the trigger is dropped. In theory, this should work for the vast majority of use cases when people make migrations. Of course there are some people that may need to put `atomic=False` in their migrations, and they may even try to do data migrations in the middle of a migration file like this. This is the problematic scenario where I would need to inform users that triggers may not actually be enabled during their migration.

It's basically either:

1. drop trigger, perform all normal migration operations (if there are any others), install any trigger updates.
2. perform all normal migration operations (if there are any others), and then update the trigger in place if it has also changed.

For 1), you get the caveat as outlined above. For 2), the ordering of when the trigger is updated may result in errors that may require the user to re-order the operations in their migration file.

I'm still leaning towards 2, but I need to think about it a bit more.

Will keep this thread updated on progress. I will visit this problem again in a week. @tomage or @simonw let me know if you have any thoughts!

**Comment by shangxiao** [comment]
@wesleykendall yeah so support for extending migrations is on my wishlist for Django. A few years back I also looked into this and came to the same conclusion that patching was the only way forward, which I was unhappy with. See https://github.com/shangxiao/django-db-constraints/blob/master/django_db_constraints/apps.py for how I did it 😕  (In my case I patched both migrations & the model meta state as I wanted to keep my db declarations alongside other model meta information)

**Comment by wesleykendall** [comment]
@shangxiao wow thanks for the tip! I will certainly look into that, especially if I can register things in the Meta options too.

**Comment by PaulGilmartin** [comment]
> I looked into the possibility of integrating django-pgtrigger directly into the migration system. It is doable, although a bit hacky as expected. I will likely have to patch the core migration loader. I was able to get an example version of code working that just created `operations.RawSQL` migration operations for installing the entire raw SQL that `django-pgtrigger` typically performs behind the scenes. I still need to evaluate how I will be able to query previous migrations to get the raw SQL of previous trigger declarations. This will allow me to more easily make reversible `RawSQL` operations. Should be doable though, just need a bit more time to explore.
> 
> Although I think I can make things work most of the time, there are particular scenarios that might be tricky to always get right. For example, if a user adds columns and references them in a trigger, we should create/update the trigger in the migration file at the end. I can't think of any examples where the trigger needs to be updated before the operation, but my hunch is that there are some options
> 
> Since ordering of the trigger installation is dependent on other operations, one easy way to alleviate these issues would be to drop/recreate the trigger in the migration file by default by surrounding all operations:
> 
> ```
> [
>     migrations.RawSQL('drop trigger...', reverse_sql='create trigger with OLD definition'),
>     migrations.RemoveField(
>        model_name='journal',
>        name='action',
>     ),
>     migrations.RawSQL('create trigger with new definition', reverse_sql='drop trigger')
> ]
> ```
> 
> If I always surround the migration operations with a drop/create for trigger changes, it seems to alleviate most issues that happen as a result of columns, table names, and other things changing.
> 
> Since Django makes migration files atomic by default, all I need to do is ensure that the initial trigger operation will take out a lock on the table so that no updates happen while the trigger is dropped. In theory, this should work for the vast majority of use cases when people make migrations. Of course there are some people that may need to put `atomic=False` in their migrations, and they may even try to do data migrations in the middle of a migration file like this. This is the problematic scenario where I would need to inform users that triggers may not actually be enabled during their migration.
> 
> It's basically either:
> 
> 1. drop trigger, perform all normal migration operations (if there are any others), install any trigger updates.
> 2. perform all normal migration operations (if there are any others), and then update the trigger in place if it has also changed.
> 
> For 1), you get the caveat as outlined above. For 2), the ordering of when the trigger is updated may result in errors that may require the user to re-order the operations in their migration file.
> 
> I'm still leaning towards 2, but I need to think about it a bit more.
> 
> Will keep this thread updated on progress. I will visit this problem again in a week. @tomage or @simonw let me know if you have any thoughts!


Continuing from the example above:
```
[
    migrations.RawSQL('drop trigger...', reverse_sql='create trigger with OLD definition'),
    migrations.RemoveField(
       model_name='journal',
       name='action',
    ),
    migrations.RawSQL('create trigger with new definition', reverse_sql='drop trigger')
]
```
What about if instead of this, the `RemoveField` class is subclassed so that the `database_forwards/database_backwards` methods apply the appropriate in-place trigger update? The resultant migration file would then simply be the regular `RemoveField` operation with the trigger handling encapsulated inside that operation. 

 I *think* the advantage of this would be that since the `RemoveField` operation is already a locking one, you could wrap the `database_forwards/database_backwards` methods in the subclassed `RemoveField` in an atomic transaction (alongside the trigger operations) at no extra cost (i.e. the user would not need to worry about declaring atomic=True at the migration file level and the fact that the operation locks anyway means there would be no functional difference by adding in another trigger operation as part of the locked transaction). This would seem to solve the main issue with solution (1) above if I understood correctly.
 
 This is all pretty theoretical and I haven't tried it in practice, but I was wondering if this idea had been explored at all?

**Comment by wesleykendall** [comment]
@shangxiao and others. Update here -

I've successfully integrated triggers with migrations in a test app. What I did was simply treat triggers as constraints and then fill in all of the methods constraints use to interact with the migration system (`deconstruct`, `create_sql`, etc).

So far things work well, but I still have a lot of testing to go. I'm also planning a way for users to still be able to register triggers on third-party models and have that work with the migration system too.

@shangxiao, I'm assuming that your app you made was before django introduced constraints, otherwise you would have been able to just make a `Constraint` class for those composite foreign keys and other things, right?
